Refactor HeaderStrip. Implement update selection functionality in machine list.

Caleb-Ellis opened the pull request:
## Done
- Split HeaderStrip into smaller components because it was starting to get large. Now it consists of:
  - AddHardwareMenu
  - TakeActionMenu
  - HeaderStripTabs
  - ActionFormWrapper (which is bare bones right now, but will house the individual form components for each action type) 
- Built the "update your selection" functionality from the angular machine list, which filters selected machines to only those that can perform the selected action.

## QA
- Select a bunch of machines.
- Open the take action menu and pick an action that not all machines can perform.
- Check that a warning shows up that asks you to update your selection.
- Click the "update your selection" link and check that the selected machines are filtered to those that can perform the action.
- Check that the submit label is formatted nicely for each action.

## Fixes
Fixes canonical-web-and-design/MAAS-squad#1793

## Screenshots
### Select action that can't be performed by all selected machines
![0 0 0 0_8400_MAAS_r_machines (14)](https://user-images.githubusercontent.com/25733845/74955980-b6671500-5405-11ea-8898-73d5ecedfca5.png)

### Error message
![0 0 0 0_8400_MAAS_r_machines (15)](https://user-images.githubusercontent.com/25733845/74955987-b9fa9c00-5405-11ea-8426-2c3ec28d268a.png)

### All machines can perform action
![0 0 0 0_8400_MAAS_r_machines (16)](https://user-images.githubusercontent.com/25733845/74955993-bc5cf600-5405-11ea-9f87-cecc0449047c.png)

huwshimi commented:
@Caleb-Ellis all the QA steps work, but after you've completed the successful action and you start checking more checkboxes the "Update your selection" warning appears again.

Caleb-Ellis commented:
@huwshimi I think I've kiiinda fixed the issue you were having? I've updated the PR to show buttons when you click on an action, where the cancel button just unsets the action from state and the submit button does nothing (at the moment).

@lilyvidenova I've added "Design needed" because of the buttons disappearing when you select an action. How's it look?

lilyvidenova commented:
Hmmm is there a reason the UI takes AGES to ./run ? It used to be done in like 2 minutes, it's been going for a solid 20 now :(

lilyvidenova commented:
@Caleb-Ellis I like how it removes the buttons! I feel like the divider line is a bit weird but we will talk about that when you're building the forms, right?

huwshimi commented:
@Caleb-Ellis thanks, I think that will work once the submit is implemented.

huwshimi commented on ui/src/app/machines/components/HeaderStrip/ActionFormWrapper/ActionFormWrapper.js:
```diff
@@ -0,0 +1,143 @@
+};
+
+ActionFormWrapper.propTypes = {
+  selectedAction: PropTypes.object, // TODO: PropType shapes for domain entities
```
Are these todos captured somewhere? If not could be good to add issues.

Caleb-Ellis replied:
Good call. Added here: https://github.com/canonical-web-and-design/maas-ui/issues/826

huwshimi commented on ui/src/app/machines/components/HeaderStrip/AddHardwareMenu/AddHardwareMenu.js:
```diff
@@ -0,0 +1,57 @@
+  );
+
+  useEffect(() => {
+    if (!navigationOptionsLoaded) {
```
Was this fetching the navigation options even if they existed in the store? I was under the impression that the saga handled checking if they were loaded. If not this might be a bug.

Caleb-Ellis replied:
Oh you're 100% right. I thought it was refetching every time but it's just calling the action "stem" (`FETCH_MACHINE` not `FETCH_MACHINE_START`). I'll remove the loaded checks.

huwshimi commented on ui/src/app/machines/components/HeaderStrip/HeaderStrip.js:
```diff
@@ -1,107 +1,29 @@
-    dispatch(machineActions.fetch());
-    dispatch(poolActions.fetch());
-  }, [dispatch]);
+    if (!machinesLoaded) {
```
Again, this should be handled by the saga, if not we should fix that.

Caleb-Ellis replied:
removed

huwshimi commented on ui/src/app/machines/components/HeaderStrip/TakeActionMenu/TakeActionMenu.js:
```diff
@@ -0,0 +1,98 @@
+  );
+
+  useEffect(() => {
+    if (!actionOptionsLoaded) {
```
Another loaded check...

Caleb-Ellis replied:
removed
